### ISSUE TITLE
Question: How to keep screen state with tab navigation destinations?

### ISSUE THREAD

**Issue by rydmike** [tomgilder/routemaster#87]
First of Tom @tomgilder, thanks for this wonderful Nav2 based package, really impressive work!

As you already know I've been trying it out a bit and have a bunch of issues, probably they are more just like use cases I have not figured out how to implement with it yet. I will be posting them more as open questions for now.

This is the first one.

## When using routemaster, how can we keep the state of the pages when using tab screens?

I made an example demonstrating the question. I used your **mobile_app** demo bundled with the package as a case example. I just added some grids to a couple of its screens (notification one and two and settings) so they become scrollable so we can easily see when the scrolling position and state is kept or lost.

As can be seen in the gif further below, it all works fine when using the bottom cupertino navigation, as that thing stores each page route in an off stage stack via its internal _TabSwitchingView. _(I will have another question on this nice, desired behavior and duplicating it for none Cupertino tabs and scaffolds in another question)_.

However, in the tab bar page route on the **Feed** page, if you scroll and switch pages, you loose the scroll position when you switch tabs. When I build tab bar views normally, the scroll position of each tab view is usually retained, which is certainly a desired feature when switching left/right in tab views. So far I was unable to duplicate that behavior with **routemaster** and the **mobile_app** example. The GIF animations below demonstrates it, by first showing that it is nicely kept on the Cupertino tabs, but not on tab bar view in Feed.

![routemaster-q1](https://user-images.githubusercontent.com/39990307/117542421-41d7c080-b021-11eb-82a4-5d92a7b06cc0.gif)

You can find the modified **mobile_app** repo with the above grids added to it here: https://github.com/rydmike/mrnavdemo

**Comment by tomgilder** [comment]
Hmm. I thought this was the standard Flutter behaviour?

Converting the `SettingsPage` to a `StatefulWidget` and adding `AutomaticKeepAliveClientMixin` keeps the scrolling position:

```dart
class SettingsPage extends StatefulWidget {
  @override
  _SettingsPageState createState() => _SettingsPageState();
}

class _SettingsPageState extends State<SettingsPage>
    with AutomaticKeepAliveClientMixin {
  @override
  bool get wantKeepAlive => true;

  @override
  Widget build(BuildContext context) {
    super.build(context);
    ...
  }
}
```

**Comment by rydmike** [comment]
Interesting, because in the case where I'm actually using it, I did not need to add the `AutomaticKeepAliveClientMixin` before. The tab view pages were stateful (well some where HookWidgets but I'm going to simplify the demo by taking them away). The tabs were still all kept alive without the mixin with my previous navigation, never saw them unloaded. However with routemaster, they were thrown out at once, and then I noticed the same in the mobile_app demo so I thought I would ask.

But yes of course the mixin totally works, thanks for the quick and obvious answer. 

It threw me off a bit since I had not needed it for the use case before, it worked fine without it. Wonder why 🤔 , but that is not so important since the keep alive mixin of course works great. 

I will soon be adding a longish post with more questions from my test with routemaster test with the Flexfold demo app. If they are this simple, then hooray! 😄 

Closing this question, thanks again!
